site/_posts/2017-10-24-api-contributors-guide.md

```diff
@@ -0,0 +1,250 @@
+* **:name:**
+    * Name of the action to be performed 
+* **:identifier:**
+    * Name of the corresponding [product feature](https://github.com/ManageIQ/manageiq/blob/master/db/fixtures/miq_product_features.yml), which is used to control permissions to resources
+* **:collection_actions:** vs **:resource_actions:**
+    * collection actions correspond to those against the collection `/api/coffee` and resource actions are those corresponding to a single resource `/api/coffee/:c_id`
+ 
+Through the above definition, the following endpoints and actions were created:
+```
+GET /api/coffee
+GET /api/coffee/:c_id
+
+POST /api/coffee
+POST /api/coffee { "action": "edit" }
+POST /api/coffee { "action": "delete" }
+POST /api/coffee/:c_id { "action": "edit" } 
+POST /api/coffee/:c_id { "action": "delete" } 
+
+DELETE /api/coffee/:c_id
+```
+### **Where are the methods?**
+Now that the routes exist, let's take care of the methods. Once the new collection has been added to the `api.yml`, the corresponding controller needs to be [created here](https://github.com/ManageIQ/manageiq-api/tree/master/app/controllers/api). The controller will look something like:
+```ruby
+module Api
+    class CoffeeController < BaseController 
+    end 
+end 
+```
+Because the `CoffeeController` inherits from the `BaseController`, a lot of what has been defined in the `api.yml` will already work :sparkles:. The [Generic Module](https://github.com/ManageIQ/manageiq-api/blob/master/app/controllers/api/base_controller/generic.rb) holds the basic CRUD methods that will be called if they are not overridden in the `CoffeeController`.
+## Custom Resource Actions
+Once basic CRUD is established, what good is seeing a coffee if you can't order one? A custom resource action is the solution for that. With the new order action, the `:resource_actions:` section will look as follows: 
+```
+        :resource_actions:
+            :get:
+            - :name: read
+              :identifier: coffee_show 
+            :post:
+            - :name: edit
+              :identifier: coffee_edit 
+            - :name: delete
+              :identifier: coffee_delete
+            - :name: order
+              :identifier: coffee_order 
+            :delete:
+            - :name: delete
+              :identifier: coffee_delete
+```
+
+Since ordering a coffee isn't a basic CRUD action handled by the magic :sparkles: generic methods, that will need to be defined in the `CoffeeController`:
+```ruby
+module Api
+    class CoffeeController < BaseController
+        def order_resource(type, id, data)
+            coffee = resource_search(id, type, collection_class(type))
+            coffee.order(data)
+        rescue => err
+            raise BadRequestError, "Coffee not ordered - #{err}"
+        end 
+    end 
+end 
+```
+The above example demonstrates a couple of the basic methods and formats that you'll see used throughout the API.
+
+* `collection_class(type)`
```

Comment:
Even though this is the way we do things everywhere....I'm wondering if this is what we still want to recommend going forward. `collection_class` adds some indirection over something that is determinate at compile time (we know what this class is), unlikely to change (the name of the model), and, if it does, easy to change. The cost of that indirection is an impediment to readability - `resource_search(id, type, Coffee)` is so much clearer IMO. Any thoughts? /cc @chrisarcand

Reply:
I agree @imtayadeway 👍

Reply:
run-time too.

Reply:
> if we refactor on our end later, we can update the doc.

But there's no refactoring necessary - this is merely reflecting how a typical vanilla CRUD endpoint would look, and there's no reason to insist that a developer use `collection_class` when there's a simpler alternative.

Reply:
Ahh ok, true.  The one thing to worry about is that new code written using above template will be inconsistent with our current code base. Personally, I'd prefer consistency.  Thanks.